Add several updates to `vim._with` (tests, granular option contexts, `env` context)

echasnovski opened the pull request:
This PR evolved into three improvements:

- *Massively* cover existing functionality with tests. Some (many) tests are marked as pending because the particular context does not work. Those tests are meant as a reference for future fixing.
- Replace `options` context with `o`, `bo`, `wo`, `go` granular contexts. This aligns with existing Lua options terminology and is more flexible than single `options`.
- Add `env` context for environment variables. As this also aligns with `vim.env`.

More details in commits' messages.

------

Resolves #29253

Alternative to #29269 and #29268

justinmk commented on runtime/lua/vim/shared.lua:
```diff
@@ -1151,34 +1154,70 @@ end
 --- @field win? integer
+--- @field wo? table<string, any>
+
+local save_set_context = function(context)
```
The name might need a bit of explanation, is this correct:
```suggestion
--- Gets option :set context.
local save_set_context = function(context)
```

Nit: "save" implies side-effects, so "get" would normally be preferrable, but `get_set_context` would be strange.

justinmk replied:
> How about `cache_and_set_context()`?

doesn't seem better than "save"

justinmk replied:
> How about `cache_and_set_context()`?

doesn't seem better than "save".  I see now that this is setting options so it does in fact have side effects, so ignore my previous comment. Just maybe a brief docstring would help.

echasnovski replied:
I resolved this by compactifying implementation so that it can be reasonably fit inside `vim._with()` body. No helper function - no problem.

echasnovski replied:
I've recently had a realization that if there is a misplaced option scope (like `{ bo = { winblend = 10 } }`), then the approach of "cache and set" inside same double loop would not properly restore already set context values (due to error during `vim.bo.winblend = 10`). So I indeed had to separate `get_context_state()` into a separate function (with brief description of what it does).

gpanders commented on runtime/lua/vim/shared.lua:
```diff
@@ -1151,34 +1154,70 @@ end
-  vim.validate('context.emsg_silent', context.emsg_silent, 'boolean', true)
-  vim.validate('context.hide', context.hide, 'boolean', true)
-  vim.validate('context.horizontal', context.horizontal, 'boolean', true)
-  vim.validate('context.keepalt', context.keepalt, 'boolean', true)
-  vim.validate('context.keepjumps', context.keepjumps, 'boolean', true)
-  vim.validate('context.keepmarks', context.keepmarks, 'boolean', true)
-  vim.validate('context.keeppatterns', context.keeppatterns, 'boolean', true)
-  vim.validate('context.lockmarks', context.lockmarks, 'boolean', true)
-  vim.validate('context.noautocmd', context.noautocmd, 'boolean', true)
-  vim.validate('context.options', context.options, 'table', true)
-  vim.validate('context.sandbox', context.sandbox, 'boolean', true)
-  vim.validate('context.silent', context.silent, 'boolean', true)
-  vim.validate('context.unsilent', context.unsilent, 'boolean', true)
-  vim.validate('context.win', context.win, 'number', true)
+  vim.validate({
+    context = { context, 'table' },
+    f = { f, 'function' },
+  })
+
+  vim.validate({
+    ['context.bo'] = { context.bo, 'table', true },
+    ['context.buf'] = { context.buf, 'number', true },
+    ['context.emsg_silent'] = { context.emsg_silent, 'boolean', true },
+    ['context.env'] = { context.env, 'boolean', true },
+    ['context.go'] = { context.go, 'table', true },
+    ['context.hide'] = { context.hide, 'boolean', true },
+    ['context.horizontal'] = { context.horizontal, 'boolean', true },
+    ['context.keepalt'] = { context.keepalt, 'boolean', true },
+    ['context.keepjumps'] = { context.keepjumps, 'boolean', true },
+    ['context.keepmarks'] = { context.keepmarks, 'boolean', true },
+    ['context.keeppatterns'] = { context.keeppatterns, 'boolean', true },
+    ['context.lockmarks'] = { context.lockmarks, 'boolean', true },
+    ['context.noautocmd'] = { context.noautocmd, 'boolean', true },
+    ['context.o'] = { context.o, 'table', true },
+    ['context.sandbox'] = { context.sandbox, 'boolean', true },
+    ['context.silent'] = { context.silent, 'boolean', true },
+    ['context.unsilent'] = { context.unsilent, 'boolean', true },
+    ['context.win'] = { context.win, 'number', true },
+    ['context.wo'] = { context.wo, 'table', true },
+  })
```
Out of curiosity, is there a reason to change this?

The single-line `vim.validate` calls are basically free. I don't think changing this is a readability win either way (there's a small argument to be made for eliminating the repeated `vim.validate`, but that's it). My main concern is about establishing patterns. Future contributors will copy-paste what they find elsewhere, and I think we should try to use the fast version of `vim.validate` wherever possible for that reason.

echasnovski replied:
My bad, it was more of an assumption that single call to `vim.validate()` is more performant than multiple. However, I did not account that those are the newly added style of vararg calls instead of creating a table every single time.

I've profiled two versions (single `vim.validate()` with table argument vs multiple with vararg) and indeed vararg version seems to be faster: around 5e-07 seconds for single `vim.validate()` and around 8e-08 seconds for vararg (time for all current 19 calls).

I'll change back to multiple vararg calls.

echasnovski committed:
test(lua): cover `vim._with()` with tests

Problem: `vim._with()` has many different use cases which are not
  covered with tests.

Solution: cover with tests. Some (many) test cases are intentionally
  marked as "pending" because they cover cases which don't work as
  expected at the moment (and fixing them requires specific knowledge of
  C codebase). Use them as a reference for future fixes.
  Also some of "can be nested" tests currently might pass only because
  the tested context doesn't work.

echasnovski committed:
feat(lua): update `vim._with` to allow more granular option contexts

Problem: with a single `context.options` there is no way for user to
  force which scope (local, global, both) is being temporarily set and
  later restored.

Solution: replace single `options` context with `bo`, `go`, `wo`, and
  `o`. Naming and implementation follows how options can be set directly
  with `vim.*` (like `vim.bo`, etc.).
  Options are set for possible target `win` or `buf` context.

dundargoc commented on runtime/lua/vim/shared.lua:
```diff
@@ -1192,29 +1221,60 @@ function vim._with(context, f)
+
+        -- Do not override already set cache and fall back to `vim.NIL` to
+        -- cache `nil` values (which needs restoring later)
+        cache[sc][name] = cache[sc][name] or vim[sc][name] or vim.NIL
```
Do I understand it correctly that e.g. if both `o` and `bo` scopes are used with different values then `o` will be used since it has higher priority?

echasnovski replied:
On the contrary, `bo` will be used as it will be set *after* setting the `o` context.

Using `cache[sc][name] = cache[sc][name] or vim[sc][name]` instead of direct `cache[sc][name] = vim[sc][name]` guards exactly from this kind of case. Without it, cache would have been set to user's value from the `o` context and later (erroneously) restored to it.

echasnovski replied:
But good question, as I realized that there was no test case specifically if same option is supplied in several contexts.

dundargoc reviewed (approved):
I haven't checked the test code but I assume they hold the standard echa quality. The option code is somewhat hard to follow, but I'm unsure if there's a good way to simplify it tbh. Maybe we can look into that later. All in all looks great. All the test cases are super valuable.

dundargoc commented on runtime/lua/vim/shared.lua:
```diff
@@ -1151,21 +1154,46 @@ end
+  vim.validate('context.env', context.env, 'table', true)
+  vim.validate('context.go', context.go, 'table', true)
   vim.validate('context.hide', context.hide, 'boolean', true)
   vim.validate('context.horizontal', context.horizontal, 'boolean', true)
```
This is an error from my part. This context does not exists in `vim._with_c` and should probably be removed for the time being.

echasnovski replied:
I removed it from code, but left pending tests just in case.

echasnovski commented on runtime/lua/vim/shared.lua:
```diff
@@ -1192,29 +1221,60 @@ function vim._with(context, f)
+    local cache = { bo = {}, env = {}, go = {}, wo = {} } ---@type table<string, any>
+
+    -- Use specific order from possibly most to least intrusive
+    for _, scope in ipairs(scope_order) do
```
I am not sure how to make it less confusing, as it is caching and setting inside for loops.

Here is an example of what it does if it would make it easier to follow. For `{ o = { winblend = 25 }, bo = { undolevels = 0 } }`, it does the following sequence:
- Process `o` context:
    - Get its "effective"/"cachable" context via `nvim_get_option_info2()`. The result is one of `bo`, `go`, or `wo`. It is `wo` for `winblend`.
    - Cache it via `cache.wo.winblend = cache.wo.winblend or vim.wo.winblend or vim.NIL`. The first `cache.wo.winblend` is there in order to write to cache only once (as the same option name can be supplied in three different `context` fields). The fallback to `vim.NIL` is there mostly for `env` context (as `vim.env` can return `nil`).
    - Cache its global value via `cache.go.winblend = cache.go.winblend or vim.go.winblend`. This is needed because later setting option might affect global values (like for `o` or `wo` contexts).
    - Set it with `vim.o.winblend = 25`.
- Process `bo` context:
    - `cache.bo.undolevels = cache.bo.undolevels or vim.bo.undolevels or vim.NIL`.
    - `cache.go.undolevels = cache.go.undolevels or vim.go.undolevels`.
    - `vim.bo.undolevels = 0`.
- Execute callback `f`.
- Restore cache:
    - `vim.bo.undolevels = cache.bo.undolevels`
    - `vim.wo.winblend = cache.wo.winblend`
    - `vim.go.undolevels = cache.go.undolevels` and `vim.go.winblend = cache.go.winblend`.

dundargoc replied:
Ah ok, that clears things up a bit. Hmm, yeah I can't come up with a simpler approach for the time being. Let's go with this 👍

echasnovski committed:
feat(lua): add `context.env` (environment variables) to `vim._with()`

justinmk commented on runtime/lua/vim/shared.lua:
```diff
@@ -1141,44 +1141,105 @@ end
+local scope_order = { 'o', 'wo', 'bo', 'go', 'env' }
+local state_restore_order = { 'bo', 'wo', 'go', 'env' }
+
+--- Get data about current state, enough to properly undo later context setting
```
```suggestion
--- Gets data about current state, enough to properly restore specified options/env/etc.
```

justinmk commented on test/functional/lua/with_spec.lua:
```diff
@@ -7,286 +7,1620 @@ local api = n.api
+    matches('context.*table', get_error("'a', function() end"))
+    matches('f.*function', get_error('{}, 1'))
+
+    local validate_context = function(bad_context, expected_type)
```
we are fairly consistent about using `assert_` prefix for tests
```suggestion
    local assert_context = function(bad_context, expected_type)
```

dundargoc commented:
This is super valuable. Thanks for championing this.

Follow-up work (not necessarily by you, I'm talking in general): ensure all the tests marked as `pending` passes. It looks like it involved primarily ensuring that `vim._with` can be nested which will need adjustments in the C code.

echasnovski commented:
> Follow-up work (not necessarily by you, I'm talking in general): ensure all the tests marked as `pending` passes. It looks like it involved primarily ensuring that `vim._with` can be nested which will need adjustments in the C code.

Not only those, I am afraid. Quite a few context fields coming from command modifiers don't seem to work:
- [`emsg_silent`](https://github.com/neovim/neovim/blob/76dd07e572935742f6dfa0d3364e262ebc047dcc/test/functional/lua/with_spec.lua#L306)
- [`hide`](https://github.com/neovim/neovim/blob/76dd07e572935742f6dfa0d3364e262ebc047dcc/test/functional/lua/with_spec.lua#L512)
- [`horizontal`](https://github.com/neovim/neovim/blob/76dd07e572935742f6dfa0d3364e262ebc047dcc/test/functional/lua/with_spec.lua#L565) (although it is not even implemented in C part)
- [`keepalt`](https://github.com/neovim/neovim/blob/76dd07e572935742f6dfa0d3364e262ebc047dcc/test/functional/lua/with_spec.lua#L591)
- [`keepjumps`](https://github.com/neovim/neovim/blob/76dd07e572935742f6dfa0d3364e262ebc047dcc/test/functional/lua/with_spec.lua#L625)
- [`keepmarks`](https://github.com/neovim/neovim/blob/76dd07e572935742f6dfa0d3364e262ebc047dcc/test/functional/lua/with_spec.lua#L654)
- [`keeppatterns`](https://github.com/neovim/neovim/blob/76dd07e572935742f6dfa0d3364e262ebc047dcc/test/functional/lua/with_spec.lua#L687)

Soo... basically most part of the flags :)
